File: examples/bun/ui.ts
/**
 * UI Helpers for METRICS monitor
 * Provides formatting, colors, and display utilities
 */

/**
 * ANSI color codes for terminal output
 */
export const colors = {
  reset: "\x1b[0m",
  bright: "\x1b[1m",
  dim: "\x1b[2m",
  cyan: "\x1b[36m",
  green: "\x1b[32m",
  yellow: "\x1b[33m",
  magenta: "\x1b[35m",
  red: "\x1b[31m",
  blue: "\x1b[34m",
  white: "\x1b[37m",
} as const;

/**
 * Format bytes to human-readable string
 * @param bytes - Number of bytes
 * @param decimals - Number of decimal places (default: 2)
 * @returns Formatted string like "1.50 GiB"
 */
export function formatBytes(bytes: number, decimals: number = 2): string {
  if (bytes === 0) return "0 B";
  
  const units = ["B", "KiB", "MiB", "GiB", "TiB", "PiB"];
  const k = 1024;
  const dm = decimals < 0 ? 0 : decimals;
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${units[i]}`;
}

/**
 * Format seconds to human-readable uptime string
 * @param seconds - Uptime in seconds
 * @returns Formatted string like "2h 30m 15s"
 */
export function formatUptime(seconds: number): string {
  const h = Math.floor(seconds / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  const s = seconds % 60;
  return `${h}h ${m}m ${s}s`;
}

/**
 * Create a progress bar
 * @param percent - Percentage (0-100)
 * @param width - Width of the bar in characters (default: 25)
 * @param filledChar - Character for filled portion (default: "█")
 * @param emptyChar - Character for empty portion (default: "░")
 * @returns Progress bar string
 */
export function createProgressBar(
  percent: number,
  width: number = 25,
  filledChar: string = "█",
  emptyChar: string = "░"
): string {
  const filled = Math.round((percent / 100) * width);
  const empty = width - filled;
  return filledChar.repeat(filled) + emptyChar.repeat(empty);
}

/**
 * Clear the terminal screen
 */
export function clearScreen(): void {
  process.stdout.write("\x1Bc");
}

/**
 * Create a colored progress bar
 * @param percent - Percentage (0-100)
 * @param color - ANSI color code for the filled portion
 * @param width - Width of the bar
 * @returns Colored progress bar string
 */
export function createColoredProgressBar(
  percent: number,
  color: string = colors.green,
  width: number = 25
): string {
  const bar = createProgressBar(percent, width);
  return `${color}${bar}${colors.reset}`;
}

/**
 * Pad a string to a specific length
 * @param str - String to pad
 * @param length - Target length
 * @param char - Padding character (default: space)
 * @returns Padded string
 */
export function padRight(str: string, length: number, char: string = " "): string {
  return str.padEnd(length, char);
}

/**
 * Format a percentage with fixed decimals
 * @param value - Percentage value
 * @param decimals - Number of decimal places
 * @returns Formatted percentage string
 */
export function formatPercent(value: number, decimals: number = 1): string {
  return `${value.toFixed(decimals)}%`;
}

/**
 * Create a header/title line
 * @param text - Header text
 * @param style - Style object with color and formatting
 * @returns Formatted header string
 */
export function createHeader(
  text: string,
  style: { color?: string; bright?: boolean } = {}
): string {
  const { color = colors.cyan, bright = true } = style;
  const prefix = bright ? colors.bright : "";
  return `${prefix}${color}${text}${colors.reset}`;
}

/**
 * Create a labeled value line
 * @param label - Label text
 * @param value - Value to display
 * @param options - Display options
 * @returns Formatted line string
 */
export function createLabeledLine(
  label: string,
  value: string,
  options: { color?: string; dim?: boolean } = {}
): string {
  const { color = colors.reset, dim = false } = options;
  const prefix = dim ? colors.dim : colors.bright;
  return `${prefix}${label}${colors.reset} ${color}${value}${colors.reset}`;
}

/**
 * Format a temperature value
 * @param temp - Temperature in Celsius
 * @param showUnit - Whether to show the unit (default: true)
 * @returns Formatted temperature string
 */
export function formatTemperature(temp: number, showUnit: boolean = true): string {
  const unit = showUnit ? "°C" : "";
  return `${temp.toFixed(1)}${unit}`;
}

/**
 * Get a color based on a percentage value
 * @param percent - Percentage value (0-100)
 * @returns ANSI color code
 */
export function getColorForPercent(percent: number): string {
  if (percent < 50) return colors.green;
  if (percent < 75) return colors.yellow;
  return colors.red;
}

/**
 * Format a number with thousands separators
 * @param num - Number to format
 * @returns Formatted number string
 */
export function formatNumber(num: number): string {
  return num.toLocaleString();
}

/**
 * Truncate a string if it exceeds a maximum length
 * @param str - String to truncate
 * @param maxLength - Maximum length
 * @param suffix - Suffix to add when truncated (default: "...")
 * @returns Truncated string
 */
export function truncate(str: string, maxLength: number, suffix: string = "..."): string {
  if (str.length <= maxLength) return str;
  return str.slice(0, maxLength - suffix.length) + suffix;
}

/**
 * Create a separator line
 * @param char - Character to use (default: "─")
 * @param length - Length of the line (default: 60)
 * @param color - Optional color for the line
 * @returns Separator string
 */
export function createSeparator(
  char: string = "─",
  length: number = 60,
  color?: string
): string {
  const line = char.repeat(length);
  return color ? `${color}${line}${colors.reset}` : line;
}

File: examples/python/ui.py
"""
UI Helpers for METRICS monitor.
Provides formatting, colors, and display utilities.
"""

import os
from typing import Optional


class Colors:
    """ANSI color codes for terminal output."""
    RESET = "\033[0m"
    BRIGHT = "\033[1m"
    DIM = "\033[2m"
    CYAN = "\033[36m"
    GREEN = "\033[32m"
    YELLOW = "\033[33m"
    MAGENTA = "\033[35m"
    RED = "\033[31m"
    BLUE = "\033[34m"
    WHITE = "\033[37m"

    # Combined styles
    BOLD_CYAN = "\033[1;36m"
    BOLD_GREEN = "\033[1;32m"
    BOLD_YELLOW = "\033[1;33m"
    BOLD_RED = "\033[1;31m"


def format_bytes(bytes_value: int, decimals: int = 2) -> str:
    """
    Format bytes to human-readable string.
    
    Args:
        bytes_value: Number of bytes.
        decimals: Number of decimal places (default: 2).
    
    Returns:
        Formatted string like "1.50 GiB".
    """
    if bytes_value == 0:
        return "0 B"
    
    units = ["B", "KiB", "MiB", "GiB", "TiB", "PiB"]
    k = 1024
    i = 0
    
    while bytes_value >= k and i < len(units) - 1:
        bytes_value /= k
        i += 1
    
    return f"{bytes_value:.{decimals}f} {units[i]}"


def format_uptime(seconds: int) -> str:
    """
    Format seconds to human-readable uptime string.
    
    Args:
        seconds: Uptime in seconds.
    
    Returns:
        Formatted string like "2h 30m 15s".
    """
    hours, remainder = divmod(seconds, 3600)
    minutes, secs = divmod(remainder, 60)
    return f"{hours}h {minutes}m {secs}s"


def create_progress_bar(
    percent: float,
    width: int = 25,
    filled_char: str = "█",
    empty_char: str = "░"
) -> str:
    """
    Create a progress bar.
    
    Args:
        percent: Percentage (0-100).
        width: Width of the bar in characters (default: 25).
        filled_char: Character for filled portion (default: "█").
        empty_char: Character for empty portion (default: "░").
    
    Returns:
        Progress bar string.
    """
    filled = int((percent / 100) * width)
    empty = width - filled
    return filled_char * filled + empty_char * empty


def create_colored_progress_bar(
    percent: float,
    color: str = Colors.GREEN,
    width: int = 25
) -> str:
    """
    Create a colored progress bar.
    
    Args:
        percent: Percentage (0-100).
        color: ANSI color code for the filled portion.
        width: Width of the bar.
    
    Returns:
        Colored progress bar string.
    """
    bar = create_progress_bar(percent, width)
    return f"{color}{bar}{Colors.RESET}"


def clear_screen() -> None:
    """Clear the terminal screen."""
    os.system('cls' if os.name == 'nt' else 'clear')


def pad_right(text: str, length: int, char: str = " ") -> str:
    """
    Pad a string to a specific length on the right.
    
    Args:
        text: String to pad.
        length: Target length.
        char: Padding character (default: space).
    
    Returns:
        Padded string.
    """
    return text.ljust(length, char)


def pad_left(text: str, length: int, char: str = " ") -> str:
    """
    Pad a string to a specific length on the left.
    
    Args:
        text: String to pad.
        length: Target length.
        char: Padding character (default: space).
    
    Returns:
        Padded string.
    """
    return text.rjust(length, char)


def format_percent(value: float, decimals: int = 1) -> str:
    """
    Format a percentage with fixed decimals.
    
    Args:
        value: Percentage value.
        decimals: Number of decimal places.
    
    Returns:
        Formatted percentage string.
    """
    return f"{value:.{decimals}f}%"


def create_header(
    text: str,
    color: str = Colors.CYAN,
    bright: bool = True
) -> str:
    """
    Create a header/title line.
    
    Args:
        text: Header text.
        color: ANSI color code.
        bright: Whether to use bright style.
    
    Returns:
        Formatted header string.
    """
    prefix = Colors.BRIGHT if bright else ""
    return f"{prefix}{color}{text}{Colors.RESET}"


def create_labeled_line(
    label: str,
    value: str,
    color: str = Colors.RESET,
    dim: bool = False
) -> str:
    """
    Create a labeled value line.
    
    Args:
        label: Label text.
        value: Value to display.
        color: ANSI color code for the value.
        dim: Whether to dim the label.
    
    Returns:
        Formatted line string.
    """
    prefix = Colors.DIM if dim else Colors.BRIGHT
    return f"{prefix}{label}{Colors.RESET} {color}{value}{Colors.RESET}"


def format_temperature(temp: float, show_unit: bool = True) -> str:
    """
    Format a temperature value.
    
    Args:
        temp: Temperature in Celsius.
        show_unit: Whether to show the unit (default: True).
    
    Returns:
        Formatted temperature string.
    """
    unit = "°C" if show_unit else ""
    return f"{temp:.1f}{unit}"


def get_color_for_percent(percent: float) -> str:
    """
    Get a color based on a percentage value.
    
    Args:
        percent: Percentage value (0-100).
    
    Returns:
        ANSI color code.
    """
    if percent < 50:
        return Colors.GREEN
    elif percent < 75:
        return Colors.YELLOW
    return Colors.RED


def format_number(num: int) -> str:
    """
    Format a number with thousands separators.
    
    Args:
        num: Number to format.
    
    Returns:
        Formatted number string.
    """
    return f"{num:,}"


def truncate(text: str, max_length: int, suffix: str = "...") -> str:
    """
    Truncate a string if it exceeds a maximum length.
    
    Args:
        text: String to truncate.
        max_length: Maximum length.
        suffix: Suffix to add when truncated (default: "...").
    
    Returns:
        Truncated string.
    """
    if len(text) <= max_length:
        return text
    return text[:max_length - len(suffix)] + suffix


def create_separator(
    char: str = "─",
    length: int = 60,
    color: Optional[str] = None
) -> str:
    """
    Create a separator line.
    
    Args:
        char: Character to use (default: "─").
        length: Length of the line (default: 60).
        color: Optional color for the line.
    
    Returns:
        Separator string.
    """
    line = char * length
    if color:
        return f"{color}{line}{Colors.RESET}"
    return line


def colorize(text: str, color: str) -> str:
    """
    Apply color to text.
    
    Args:
        text: Text to colorize.
        color: ANSI color code.
    
    Returns:
        Colorized text string.
    """
    return f"{color}{text}{Colors.RESET}"
